### PR TITLE
fix for flaky-tests

### PR DESCRIPTION
**Pull request by TimMaltGermany**
@noorul @hntd187 I just set up this branch to address the flacky test.

The latest failure on master is caused by too few job slots: 
java.lang.AssertionError: assertion failed: expected class spark.jobserver.CommonMessages$JobResult, found class spark.jobserver.CommonMessages$NoJobSlotsAvailable

I think that can be easily fixed. 

The other issues reported by @hntd187 are:
The tests in spark.jobserver.auth.WebApiWithAuthenticationSpec fail sometimes for no obvious reason. Given how they are written it seems really strange they fail, but one is ignored and the other 2 are not. Could we make a decision on whether or not these tests should remain or set to ignore for the time being? They seem to cover a rather edge case situation.

### CODE REVIEW

**Comment by TimMaltGermany** [comment]
@velvia Do you have any idea why there might not be any job slots left? This appears to be some leftover from earlier tests.

**Comment by velvia** [comment]
Not too sure.  Which test is this again?

-Evan
"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead

**Comment by TimMaltGermany** [comment]
It is NamedObjectsJobSpec.describe("NamedObjects (DataFrame + RDD)")
Interesting is that the test fails in line 101 which is the 9th job that is started by this test, but maybe that is just a coincidence. Could it be that the previous SQL jobs do not properly finish?

**Comment by TimMaltGermany** [comment]
I increased the max number of jobs per context to 8 for this test. Of course it might be just luck that it did not fail there this time. 
Now it fails at:
[info] - jars should not allow user with valid authorization when timeout *** FAILED ***

[info]   200 OK was not equal to 500 Internal Server Error (WebApiWithAuthenticationSpec.scala:400)
which is the issue reported by @hntd187

**Comment by TimMaltGermany** [comment]
uh, now it fails in JobManagerSpec at:
- should be able to start a job with job jar dependencies and return results *** FAILED ***
...
[info]   at spark.jobserver.JobManagerSpec$$anonfun$2$$anonfun$apply$mcV$sp$22.apply$mcV$sp(JobManagerSpec.scala:220)

**Comment by hntd187** [comment]
@TimMaltGermany I believe the JobManagerSpec test is something that @velvia is disabling in prep for the new Job API, so it's probably okay if you set it to ignore as well.

**Comment by TimMaltGermany** [comment]
Is there a way to manually trigger another build? 
I would like to run a few more tests to make sure that these changes really had an effect.

**Comment by velvia** [comment]
@TimMaltGermany I can restart the build again I think.   However another thing to try is to temporary disable the other tests to see if running fewer tests in that suite really affects it.

Also would be great if you could explain what the authentication test is trying to do underneath the hood, as I don't understand it, hopefully that could give more insight as to why it might fail?

**Comment by velvia** [comment]
Restarted.  Maybe it's a good idea to collapse all the commits and rebase / push -f afterwards.

**Comment by TimMaltGermany** [comment]
The NamedObjectsJobSpec test still fails occasionally due to 'NoJobSlotsAvailable'. Turns out that I didn't change the parameter at all. I will have another look. These tests really shouldn't influence each other.

**Comment by TimMaltGermany** [comment]
@velvia I will consolidate the check-ins once this is done

I have changed a few things around just to see whether that has any effect:

1.  HiveContextFactory in HiveJobSpec needs to stop the SparkContext if it fails to create the Hive context. We might want to add this check to 'spark.jobserver.HiveContextFactory' as well, just in case.

2. NamedObjectsSpec only creates its SparkContext when run and only once.

3. JobSpecBaseBase.afterAll uses TestKit.shutdownActorSystem(system) instead of the ooyala version. It also only shuts down the manager in afterAll (as opposed to after each method).

File: job-server-extras/test/spark.jobserver/HiveJobSpec.scala [hunk 1]
```diff
@@ -9,8 +10,15 @@ import spark.jobserver.context.{HiveContextLike, HiveContextFactory}
...
-    new TestHiveContext(new SparkContext(conf)) with HiveContextLike
+  override protected def contextFactory(conf: SparkConf): C = {
+    val sc = new SparkContext(conf)
+    Try(new TestHiveContext(sc) with HiveContextLike) match {
+      case Success(hc) => hc
+      case Failure(e) =>
+        sc.stop
+        throw e
+    }
+  }
```
Comment (TimMaltGermany) [inline job-server-extras/test/spark.jobserver/HiveJobSpec.scala]: This was required as this test fails on my Windows machine and it did not clean-up after itself.

**Comment by TimMaltGermany** [comment]
I give up. Now it fails for JobManagerSpec when trying to retrieve the extra jar path ('getExtrasJarPath').
Is this new or have any of you seen this before as well?

**Comment by velvia** [comment]
Is this the same test that I disabled in the new job API PR?
That one fails due to the extras jar path.  If you run the job-server-extras/test first, then it works.

**Comment by TimMaltGermany** [comment]
@velvia I guess so. Once again, this failure only happens on travis. It only failed in one of 4 tests - so this also seems to be a timing problem. Could you please trigger the build again so that we see what happens? Thanks.

**Comment by hntd187** [comment]
@TimMaltGermany Restarted

**Comment by hntd187** [comment]
LGTM @TimMaltGermany any chance you can squash some of the commits down?

**Comment by TimMaltGermany** [comment]
Done - seems like I also incorporated @noorul 's check-ins - hope that is OK?

**Comment by TimMaltGermany** [comment]
It might be cleaner and easier to just create a new Branch/PR.

**Comment by noorul** [comment]
I think it will be better to rebase and squash

**Comment by TimMaltGermany** [comment]
Well, I did rebase but unfortunately including your last check-ins. Is there a way to undo that?

**Comment by noorul** [comment]
Maybe it will be ideal to have separate branch and pr

**Comment by TimMaltGermany** [comment]
ok, sorry about the confusion. Will do it tomorrow.

**Comment by TimMaltGermany** [comment]
created PR #505 instead
